pim-tool-back/src/main/java/vn/elca/training/service/impl/ProjectServiceImpl.java

```diff
@@ -1,30 +1,193 @@
+                }
+            }
+        }
+        //status string must be not null and is one of the values in Project.Status
+
+        if (conditions != null && !StringUtils.isBlank(status) && Project.Status.contains(status)) {
+            conditions = conditions.and(QProject.project.status.eq(Project.Status.valueOf(status)));
+        }
+        // Build sort request by project number
+        Sort sort = Sort.by("projectNumber").ascending();
+        // Build the pagination request
+        Pageable pageable = PageRequest.of(page, limit, sort);
+        // Retrieve the projects from the database
+        return projectRepository.findAll(conditions, pageable).map(mapper::projectToProjectDto);
+    }
+
     @Override
-    public List<Project> findAll() {
-        return projectRepository.findAll();
+    public ProjectDto findById(long id) {
+        return mapper.projectToProjectDto(projectRepository.findById(id).orElseThrow(ProjectNotFoundException::new));
+    }
+
+    @Override
+    public ProjectDto findProjectByNumber(Integer number) {
+        return mapper.projectToProjectDto(projectRepository.findByProjectNumber(number).orElseThrow());
+    }
+
+    @Override
+    public boolean delete(long id) {
+        return false;
+    }
+
+    @Override
+    public void deleteByProjectNumber(Integer projectNumber) throws ProjectNotInNewStatusException {
+        Project project = projectRepository.findByProjectNumber(projectNumber).orElseThrow(ProjectNotFoundException::new);
+        if (project.getStatus() != Project.Status.NEW) {
+            //just delete the project if it is not in NEW status
+            throw new ProjectNotInNewStatusException("Project is not in NEW status");
+        }
+        projectRepository.deleteByProjectNumber(projectNumber);
     }
 
     @Override
     public long count() {
         return projectRepository.count();
     }
+
+    @Override
+    public Page<ProjectDto> findAllOrderByProjectNumber(int page, int limit) {
+        // Build the sorting request
+        Sort sort = Sort.by("projectNumber").ascending();
+        // Build the pagination and sorting request
+        Pageable pageableAndSorting = PageRequest.of(page, limit, sort);
+        // Retrieve the projects from the database
+        Page<Project> projects = projectRepository.findAll(pageableAndSorting);
+        return projects.map(mapper::projectToProjectDto);
+    }
+
+    private void attachEmployeesToProject(Project project, String members) {
+        if (!StringUtils.isBlank(members)) {// if members is not empty
+            List<String> visas = List.of(members.split(","));
+            if (employeeRepository.existsByVisaIn(visas)) {
+                project.setEmployees(new HashSet<>(employeeRepository.findByVisaIn(visas)));
+            } else {//throw exception that project's members are not valid
+                throw new InvalidProjectMembersException("Project's members are not valid.");
```

Comment:
[S] Which members are invalid?

Reply:
when the user enters the project's visa with the textfield hint it is almost impossible that a visa among the project's visas is not valid, in case I haven't suggested it when the user input it then I will assume that there is a possibility that the employee does not exist or is invalid